### PR TITLE
(PUP-4000) Make optional Struct value imply optional key

### PR DESCRIPTION
**Pull request by thallgren**
This commit changes the TypeCalculator so that a Struct with optional
values becomes assignable from a Struct where the keys corresponding
to those values are missing. It also changes the instance_of logic so
that a Hash missing such keys is considered an instance of the struct.

### CODE REVIEW

File: lib/puppet/pops/types/type_calculator.rb [hunk 1]
```diff
@@ -452,9 +452,18 @@ def instance_of_PTupleType(t, o)
 
   def instance_of_PStructType(t, o)
     return false unless o.is_a?(Hash)
-    h = t.hashed_elements
```
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: this gives a cached hash of key to value in the struct - why not use it ?

File: lib/puppet/pops/types/type_calculator.rb [hunk 2]
```diff
@@ -452,9 +452,18 @@ def instance_of_PTupleType(t, o)
...
-    # all keys must be present and have a value (even if nil/undef)
-    (o.keys - h.keys).empty? && h.all? { |k,v| instance_of(v, o[k]) }
```
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: I don't get why this did not work. The first part (o.keys - h.keys) should be empty if o has fewer keys than h. The second part where instance_of(v, o[k]) should end up checking if nil is instance of `Optional[T]` which it should be.

File: lib/puppet/pops/types/type_calculator.rb [hunk 3]
```diff
@@ -452,9 +452,18 @@ def instance_of_PTupleType(t, o)
...
+    match_count = 0
+    ok = t.elements.reduce(true) do |memo,e|
+      break false unless memo
+      v = o[e.name]
+      if v.nil?
+        e.type.is_a?(Types::POptionalType)
```
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: This is not correct - there are other types that accept Optional, such as a Variant that includes an Optional, or Data.

File: lib/puppet/pops/types/type_calculator.rb [hunk 4]
```diff
@@ -1192,22 +1201,34 @@ def tuple_entry_at(tuple_t, from, to, index)
   # @api private
   #
   def assignable_PStructType(t, t2)
-    return true if t == t2 || t.elements.empty? && (t2.is_a?(Types::PHashType))
-    h = t.hashed_elements
+    els = t.elements
```
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: 'els' as short for elements ? (not such a great name, looks like misspelled 'else'

File: lib/puppet/pops/types/type_calculator.rb [hunk 5]
```diff
@@ -1192,22 +1201,34 @@ def tuple_entry_at(tuple_t, from, to, index)
...
+    return true if t == t2 || els.empty? && (t2.is_a?(Types::PHashType))
     if t2.is_a?(Types::PStructType)
       h2 = t2.hashed_elements
-      h.size == h2.size && h.all? {|k, v| assignable?(v, h2[k]) }
+      match_count = 0
+      ok = els.reduce(true) do |memo, e|
+        break false unless memo
+        t2e = h2[e.name]
+        if t2e.nil?
+          e.type.is_a?(Types::POptionalType)
```
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: same problem here - not a good enough check.

File: lib/puppet/pops/types/type_calculator.rb [hunk 6]
```diff
@@ -1192,22 +1201,34 @@ def tuple_entry_at(tuple_t, from, to, index)
   # @api private
   #
   def assignable_PStructType(t, t2)
-    return true if t == t2 || t.elements.empty? && (t2.is_a?(Types::PHashType))
-    h = t.hashed_elements
+    els = t.elements
+    return true if t == t2 || els.empty? && (t2.is_a?(Types::PHashType))
```
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: This is odd - accepting any Hash if empty is fine, since if the t Struct has no elements it means any struct, and thus any hashtype, but that is also true of any struct type.

File: lib/puppet/pops/types/type_calculator.rb [hunk 7]
```diff
@@ -1192,22 +1201,34 @@ def tuple_entry_at(tuple_t, from, to, index)
...
     if t2.is_a?(Types::PStructType)
       h2 = t2.hashed_elements
-      h.size == h2.size && h.all? {|k, v| assignable?(v, h2[k]) }
```
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: this line should be changed to subtract keys, and then check assignability of either the corresponding type in the t2 struct, or PNilType if that key did not have a type.

File: lib/puppet/pops/types/type_calculator.rb [hunk 8]
```diff
@@ -1192,22 +1201,34 @@ def tuple_entry_at(tuple_t, from, to, index)
...
+          match_count += 1
+          assignable?(e.type, t2e)
+        end
+      end
+      ok && match_count == h2.size
     elsif t2.is_a?(Types::PHashType)
       size_t2 = t2.size_type || @collection_default_size_t
       size_t = Types::PIntegerType.new
-      size_t.from = size_t.to = h.size
+      size_t.from = els.reduce(0) { |c, e| e.type.is_a?(Types::POptionalType) ? c : c + 1 }
```
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: this is not good enough since other types that Optional may accept Optional

**Comment by hlindberg** [comment]
Sorry, back to the drawing board on this one.

**Comment by thallgren** [comment]
Repoening this since I think the approach taken has some merits that I want to discuss.

File: lib/puppet/pops/types/type_calculator.rb (hunk 1 above)
Comment (thallgren) [inline lib/puppet/pops/types/type_calculator.rb]: Because a Struct may well be very short-lived (a Hash is inferred to a Struct) and the cache in that case causes more harm than good. Reverse the question, why use it when it's not needed here?

File: lib/puppet/pops/types/type_calculator.rb (hunk 2 above)
Comment (thallgren) [inline lib/puppet/pops/types/type_calculator.rb]: No need to change unless you want to limit the use of copies in the code. The new approach creates no hash, nor does it cache it. Only one side needs a hash to maximize efficiency during matching.

File: lib/puppet/pops/types/type_calculator.rb (hunk 6 above)
Comment (thallgren) [inline lib/puppet/pops/types/type_calculator.rb]: I removed the special test since even if t2 is a Hash, it's key_type needs to be assignable to the Struct key_type (e.g. String). The normal test is quick enough anyway.

File: lib/puppet/pops/types/type_calculator.rb (hunk 2 above)
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: oh, so this was an optimization - not a fix. I am fine with that, but put it in a separate commit please.

Comment (thallgren) [inline lib/puppet/pops/types/type_calculator.rb]: To be honest, it didn't start as an optimization. I didn't realize that nil was assignable to Optional, hence the change. I can move it into a separate commit, but doesn't that also imply creating a separate issue and a separate PR? If so, then let's just forget about it. It's not worth all that.

File: lib/puppet/pops/types/type_calculator.rb [hunk 9]
```diff
@@ -1192,27 +1200,44 @@ def tuple_entry_at(tuple_t, from, to, index)
...
+        break false unless memo
+        t2e = h2[e.name]
+        if t2e.nil?
+          is_type_optional?(e.type)
+        else
+          match_count += 1
+          assignable?(e.type, t2e)
+        end
+      end
+      ok && match_count == h2.size
```
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: same applies to this optimization

Comment (thallgren) [inline lib/puppet/pops/types/type_calculator.rb]: Although this is not an optimization. The old code tested if {{nil}} is a {{Type}} that is assignable to an {{Optional}}. I assume that isn't the case since nil is not a type at all.

File: lib/puppet/pops/types/type_calculator.rb [hunk 10]
```diff
@@ -1192,27 +1200,44 @@ def tuple_entry_at(tuple_t, from, to, index)
...
+        (size_t2.to == 0 || assignable?(@non_empty_string_t, t2.key_type)) &&
+          elements.all? {|e| assignable?(e.type, element_type) }
     else
       false
     end
   end
 
+  # api private
+  #
+  def is_type_optional?(t)
```
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: Data and Any also accepts nil. Not a fan of special methods for checking type. We have one or so already. I would rather it went in the other direction with fewer ways to check if something has a particular type.

Comment (thallgren) [inline lib/puppet/pops/types/type_calculator.rb]: Accepting nil is very different from not supplying a value at all. Please note that the is_type_optional? is only called when there is a value (i.e. when _hash.include?_ responds *true*). It is not called for existing keys with nil values.

File: lib/puppet/pops/types/type_calculator.rb (hunk 9 above)
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: That was the real problem then? instead of using h2[k] directly, it should have used a PNilType. The rest is the same change as above that does not use all?

Comment (thallgren) [inline lib/puppet/pops/types/type_calculator.rb]: Again, not providing a value at all is quite different from providing a nil value. Don't you agree with that?

**Comment by hlindberg** [comment]
I want to see the problem fixed clearly without any optimizations. We have tickets to later measure and optimize - lets to such things then.

**Comment by thallgren** [comment]
I'd argue that the "optimization" is a natural part of this PR since it was made possible by the introduction of the _is_optional_type?_ method. It also makes the _instance_of_PStructType_ more similar to the _assignable_PStructType_ since they use the same approach.

**Comment by hlindberg** [comment]
I don't like the introduction of is_optional_type - unnecessary specialized check. It is hard enough to read and reason about the type system methods without optimizations - if we can use straight forward code like: 
```
(o.keys - h.keys).empty? && h.all? { |k,v| instance_of(v, o[k]) }
```
instead of over > 10 lines of code with conditionals and counters and stuff - then I very much prefer
the former.

I like the tests though :-)

File: lib/puppet/pops/types/type_calculator.rb (hunk 2 above)
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: yes, please keep the original version - there was nothing wrong with it

File: lib/puppet/pops/types/type_calculator.rb (hunk 9 above)
Comment (thallgren) [inline lib/puppet/pops/types/type_calculator.rb]: This test demostrates where just using ```PNilType``` (and also introducing the necessary but slow ```(h2.keys - h.keys).empty?```) would go wrong:

      it 'should reject matching structs with less elements when unmatched elements are data' do
        struct1 = struct_t({'a'=>Integer, 'b'=>Integer, 'c'=>data_t})
        struct2 = struct_t({'a'=>Integer, 'b'=>Integer})
        calculator.assignable?(struct1, struct2).should == false
      end

Would you consider that test invalid?

Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: In general that is different - but not here. If the StructType T says key 'a' is Optional, then it is assignable with a StructType that does not have an 'a' at all, or a StructType that has a type for its 'a' that is assignable to T's 'a' type.

File: lib/puppet/pops/types/type_calculator.rb (hunk 10 above)
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: It is specified that an Optional entry in a Struct means that the key may be missing, or it may have a nil (i.e. undef) value.

**Comment by thallgren** [comment]
Code simplified and tests retained :-)

File: lib/puppet/pops/types/type_calculator.rb [hunk 11]
```diff
@@ -1192,22 +1192,23 @@ def tuple_entry_at(tuple_t, from, to, index)
...
+      h = t.hashed_elements
       h2 = t2.hashed_elements
-      h.size == h2.size && h.all? {|k, v| assignable?(v, h2[k]) }
+      (h2.keys - h.keys).empty? && h.all? {|k, v| v2 = h2[k]; assignable?(v, v2.nil? ? @nil_t : v2) }
     elsif t2.is_a?(Types::PHashType)
       size_t2 = t2.size_type || @collection_default_size_t
       size_t = Types::PIntegerType.new
-      size_t.from = size_t.to = h.size
+      elements = t.elements
+      size_t.from = elements.reduce(0) { |c, e| assignable?(e.type, @nil_t) ? c : c + 1 }
```
Comment (hlindberg) [inline lib/puppet/pops/types/type_calculator.rb]: this can be written
```
size_t.from = elements.count {|e| !assignable?(e.type @nil_t) }
```
That is "the count of required keys"
